### PR TITLE
Add document which describes agent architecture and core code abstraction

### PR DESCRIPTION
**Pull request by Kami**
This pull request adds a document which describes agent architecture and core code abstractions.

It's mostly based on the notes I took yesterday during the brain dump from @czerwingithub.

I decided to add it here instead of internal wiki for 2 main reasons:

1. This document contains no proprietary or sensitive information. It just documents the concepts and components which are open-source and part of this repo.

2. I believe documentation should live as close to the source as possible. In this case, that's this repo.

I have bad experience with "central" wikis or similar. Usually when such information is not co-located with the code, it's a much higher chance it will get out of date.

And I would argue that out of date / wrong documentation is better than no documentation.

That's also the reason why this brain-dump mostly focused on higher level abstractions which are less likely to change. For low level implementation details, the code still is best source of truth.

### CODE REVIEW

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+* Agent supports Linux and Windows
+* Platform specific functionality is available in ``platform_*.py`` files
+
+### 2. Copy Manager ([copy_manager.py](https://github.com/scalyr/scalyr-agent-2/blob/master/scalyr_agent/copy_manager.py))
+
+* It's responsible for reading the log content from disk and sending it to the Scalyr API
+* It's told what it should be reading / copying.
+  * It get's a list of log files which it should be copying / uploading to Scalyr + the
+    corresponding log file settings (what parser should be used, any reduction or samplying rules,
```
Comment (czerwingithub) [inline docs/architecture.md]: s/reduction/redaction/

Comment (czerwingithub) [inline docs/architecture.md]: s/samplying/sampling/

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+* Reads bytes out of the log file and inserts it into the API request.
+* It also manages check points.
+* Check point represents up to which point in file we have copied bytes so far. To take log
+  rotation, file moving and other possible changes into account, it relies on the ``LogFileIterator``
+  abstraction.
+
+#### LogFileIterator ([log_processing.py](https://github.com/scalyr/scalyr-agent-2/blob/master/scalyr_agent/log_processing.py))
+
+* It's responsible to watch a single file path and give an illusion that you are iterating over
+  all the bytes in a log file even if that log file has been rotated.
```
Comment (czerwingithub) [inline docs/architecture.md]: Here I would say "iterating over all log lines in a log file".  I just want to make the point the abstraction iterates at the line level.

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    enabled for the main scalyr-agent-2 process
+* A thread is spawned for each monitor and MonitorsManager manages those threads
+* Monitors either discover log files or produce new files on disk
+  * This way it can rely on the same abstractions which are used for sending log file content to
+    Scalyr API also for sending data / metrics produced by those monitors.
+* Once it writes data to disk, it's responsibility ends. After that, copy manager is responsible for
+  picking up that content and sending it to Scalyr API.
+* For safety reasons, there are rate limits put in place for each monitor writes. Each write
+  represents data which is sent to the Scalyr API and that's data end user needs to pay for. To
+  avoid or bad plugins which could result in a lot of data being produced and written to disk, rate
```
Comment (czerwingithub) [inline docs/architecture.md]: Remove the `or`

**Comment by Kami** [comment]
I believe I've addressed all the feedback in 6cecb5443017eb405eeae42897b061255770d7d4.

File: docs/architecture.md [hunk 4]
```diff
@@ -70,7 +70,7 @@ starting a thread, stopping a thread and asking thread for a status.
 #### LogFileIterator ([log_processing.py](https://github.com/scalyr/scalyr-agent-2/blob/master/scalyr_agent/log_processing.py))
 
 * It's responsible to watch a single file path and give an illusion that you are iterating over
-  all the bytes in a log file even if that log file has been rotated.
+  all long lines in a log file even if that log file has been rotated.
```
Comment (czerwingithub) [inline docs/architecture.md]: Just delete `long`.  I think you meant `log` but it's probably best to leave the word out entirely.

**Pull request merged**